### ISSUE TITLE
TS5055: Typescript includes files in "dist" folder for compilation when `rootDir` is not `./` but `./src` even after explicitly adding "dist" to `exclude`

### ISSUE THREAD

**Issue by calebpitan** [microsoft/TypeScript#58090]
### 🔎 Search Terms

TS5055, rootDir, outDir, exclude

### 🕗 Version & Regression Information

TypeScript 5.4.4

### ⏯ Playground Link

https://github.com/calebpitan/zne-repro

_No response_

### 💻 Code

```bash
// Your code here

yarn install

cd ./libs/common

yarn tsc --build tsconfig.json
```

This fails with the TS5055 error and since it is the base of every other dependent workspace/lib package build for the others fail too, except I `rimraf dist` which works fine if I just want to do a one time build, but doesn't if I have to build-watch.

**NOTE**: Initial build always works fine, subsequent builds are the ones that fail.

### 🙁 Actual behavior

Includes output files (mostly declaration output files) as part of input even after explicitly excluding them. This doesn't happen when I change `rootDir` from `./src` to `./` (but unfortunately this creates an unwanted effect of emitting output to `dist/src` instead of just `dist` so I'd like to retain `rootDir` as `./src` rather than `./`)

### 🙂 Expected behavior

Should not include output files in input files especially after explicitly excluding them

### Additional information about the issue

**tsconfig.json**

```jsonc
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "./dist",
    "rootDir": "./src",
    "baseUrl": "."
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist"]
}
```

**tsconfig.base.json**

```jsonc
{
  "compilerOptions": {
    "module": "commonjs",
    "moduleResolution": "node",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "target": "ES2022",
    "sourceMap": true,
    "outDir": "./dist",
    "baseUrl": "./",
    "isolatedModules": true,
    "incremental": true,
    "skipLibCheck": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": false,
    "noImplicitAny": false,
    "strictBindCallApply": false,
    "forceConsistentCasingInFileNames": false,
    "noFallthroughCasesInSwitch": false,
    "resolveJsonModule": true,
    "allowJs": false,
    "jsx": "react",
    "typeRoots": ["./@types", "node_modules/@types"],
    "paths": {
      "@/module/*": ["./apps/customer/src/module/v1/*"]
    }
  },
  "watchOptions": {
    "excludeDirectories": ["*/**/node_modules", "*/**/dist", "libs/**/dist"]
  }
}
```

First compilation runs smoothly, subsequent compilations error with:

```
error TS5055: Cannot write file '/Users/calebpitan/source/work/company/project/libs/customer/dist/constant.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '/Users/calebpitan/source/work/company/project/libs/customer/dist/index.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '/Users/calebpitan/source/work/company/project/libs/customer/dist/prisma.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '/Users/calebpitan/source/work/company/project/libs/customer/dist/prisma.enum.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '/Users/calebpitan/source/work/company/project/libs/customer/dist/prisma.module.d.ts' because it would overwrite input file.
```

**Comment by MartinJohns** [comment]
Just FYI, `exclude` is only used to exclude files that are included by `include`. You can't explicitly exclude files from compilation. If they're imported in your code, they'll be used for compilation.

**Comment by calebpitan** [comment]
Hi @RyanCavanaugh. I managed to set up something you can use to repro.

https://github.com/calebpitan/zne-repro

```bash
yarn install

cd ./libs/common

yarn tsc --build tsconfig.json
```

This fails with the TS5055 error and since it is the base of every other dependent workspace/lib package build for the others fail too, except I `rimraf dist` which works fine if I just want to do a one time build, but doesn't if I have to build-watch.

Thanks. I'd be expecting your feedback!

**Comment by calebpitan** [comment]
> Just FYI, `exclude` is only used to exclude files that are included by `include`. You can't explicitly exclude files from compilation. If they're imported in your code, they'll be used for compilation.

@MartinJohns  Do declaration files really get imported, because it's only the declaration files causing this issue?

**Comment by MartinJohns** [comment]
Yes, declaration files are part of the compilation. They provide type information for JavaScript code.

Have you tried the flag I suggested to see why the file is being included?

**Comment by calebpitan** [comment]
Yes, I did. It’s a very long list of various files including node_modules source paths, it so difficult to track and analyze anything from it

**Comment by calebpitan** [comment]
Hi @RyanCavanaugh @MartinJohns , I've been able to resolve this. It had to do with `typeRoots` in base config which is extended by other workspaces's config. So in a way one of the workspaces re-references itself since some of the types under the `./@types` folder imports a type reference from that workspace to use as declaration merging for express.

I moved the `typeRoots` from base config which is used for extension to the root `tsconfig.json`.

> We'd need a concrete repro (what the files actually) to confirm, but ~100% of the time this is a misconfiguration.

My configuration was broken, but one thing that has stayed unbroken is your "~100% of the time" streak :)

Cheers 🥂
